### DISCUSSION TITLE
Replace PDK Components and CrossSections in cell decorator

### DISCUSSION THREAD

**Discussion by thomasdorch** [gdsfactory/gdsfactory#409] [Ideas]
It would be nice to not have to replace all default components/cross_sections with string arguments for them to be overwritten by a registered component/cross_section in an active PDK. I added some functionality to the cell decorator in [this branch](https://github.com/thomasdorch/gdsfactory/tree/use_cell_for_pdk_replacements). 

The idea is that we can keep the default arguments to components like `straight` as the CrossSection `strip`, and @cell will just look in the active PDK for a CrossSection with the same name and add that as a kwarg rather than setting the default to a string containing `"straight"`. Check out the mzi component for reference on how this works.
This change only adds kwargs, so it shouldn't break what's happening in the `pdk.py` methods.

This might not be what everyone else has in mind for how the PDK gets handled so I'm not putting it up as a PR just yet but I'd like to see if anyone has thoughts on this. I also think it can be used to handle the labels as mentioned in #404 and #405, because matching the name of the kwarg rather than the argument itself is straightforward, so anytime a `layer_label`, `instance_label` or `port_layer`, etc. kwarg is used, @cell can try to pull a replacement from the active pdk if one is defined. 

What do you guys think? @tvt173 @joamatab @Jan-David-Black

**Comment by joamatab** [comment]
Hi Thomas,

I like the idea of replacing cross-section functions by strings 

the cell magic seems a bit too magic to explain :) maybe just adding a LayerSpec that can be a tuple or string, and writing a pdk.get_layer would be more clear

what do you think?

**Reply by jan-david-fischbach** [reply]
Hi Joaquin, Hi Thomas,
when referring to replacing the cross-section functions by strings, do you also consider updating the default values of all gf.components, which is currently only partially in place from what I see? Or would a PDK have to override all these using partial (which might lead to unnoticed errors, as some components work out of the box with string cross sections)? If we can bring this to a consistent state, and do a similar thing for Layer resolution from strings as suggested, I would think this would be the least error prone and the most transparent way to handle this.
Regarding Thomas approach I really like the fact, that it works out of the box across the board, but I am not sure how well name collisions between for example cross sections and cells are handled by the code provided in the branch. As far as I see, a collision would be resolved by always choosing the cross section over the cell 
```python
attr = (
              "cross_sections"
              if param_name in pdk.cross_sections.keys()
              else "cells"
              if param_name in pdk.cells.keys()
              else "containers"
              if param_name in pdk.containers.keys()
              else None
          )
```
please correct me if I am wrong. I do not necessarily find that desired behavior.

**Reply by tvt173** [reply]
in general, I agree... the less magic the better (and I think the [zen of python](https://peps.python.org/pep-0020/) also agrees😄) 

i think if we start to use a LayerSpec, as @joamatab suggests, it will be a big refactor at the beginning, but once we get it in, as it was with components, it should "just work" in most places, and is still a fairly straightforward process to follow when you need to call `get_layer()` explicitly

i'm hesitant to let the decorator magically convert the arguments. the way it is now, though it requires an extra line of code in some places, I think is straightforward and plays nicely with most IDEs

i also think that any default argument value for a Component or CrossSection that is intended to be resolved dynamically *should* be defined as a string. to convert a reference to a function to a different function entirely would be [very surprising](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) to most programmers, but it is very clear that a string value must be dynamically resolved somehow

what's your main frustration now? if you don't like relying on freely typed strings, maybe you could do something like `component=my_component.__name__`?

**Reply by thomasdorch** [reply]
The primary reason I thought this would be useful was to be able to keep the connections between functions clear. As far as I can tell (and I'm not an expert in IDEs by any means) using a string as the argument for `cross_section` means we have to go find the right module, and track down the specific `cross_section` declaration, rather than being able to jump directly to the function using some shortcut or hovering and seeing how it's defined. 

That being said, I think @tvt173's justification for _not_ using the cell decorator for this is hard to argue with -- it would likely be confusing to have a decorator replacing arguments and functions without explicitly being asked to. Additionally, dealing with nested partials using my method is really annoying.

I really like the idea of using the LayerSpec going forward, and it will likely be able to do exactly what this is intended to do, in a more predictable way. Especially since pretty much every time I need to override the default `cross_sections`, I'm only replacing the layer arguments.

Based on what was suggested by @bradley-w-snyder in #405, it would be a lot nicer to have `cross_section=DEFAULT.strip`, which would link to a default class that contains something like `"strip"=strip`, so there is still some connection between the strip function and the string argument.

**Reply by tvt173** [reply]
hi @thomasdorch ,
the challenge with having a pattern like 
```python
def my_function(cross_section=DEFAULT.strip):
```

is that `DEFAULT.strip` will be evaluated at import time... so if it were pointing to i.e. `(1, 0)` it would be difficult to do well in a way that will be dynamic and flexible, but still easy enough to understand...

what might be a reasonable option though, would be to have something like

```python
## ADD TO GDSFACTORY
from pydantic import BaseModel
from gdsfactory.types import CrossSectionSpec
import gdsfactory as gf


class DefaultCrossSections(BaseModel):
    strip: CrossSectionSpec = "strip"
    rib: CrossSectionSpec = "rib"
    

DEFAULT_CS_NAMES = DefaultCrossSections()
...

## YOUR SCRIPT
...

def my_function(cross_section: CrossSectionSpec = DEFAULT_CS_NAMES.strip):
    pass


# YOUR PDK DEFINITION
...

MY_CROSS_SECTIONS = DefaultCrossSections(
    strip=gf.cross_section.strip,
    rib=gf.cross_section.rib,
)

PDK = gf.pdk.Pdk(name='my_pdk',
                 cross_sections=MY_CROSS_SECTIONS.dict(),
                 )
```

note that we still effectively use strings as the default argument values, but they are statically defined. and you are able to create a static mapping to your cross sections, like you requested, and it all works as intended.

now that we have `LayerSpec` defined, this same pattern should work for layers, cross sections, and components.

**Reply by thomasdorch** [reply]
Thanks, I hadn't considered that issue with importing! 

I really like this setup, particularly because it would make explicit which cross sections need to be overridden by a PDK without adding too much overhead. The same goes for the components and layers, and might even be a good way to implement a default routing method as we discussed in [this thread](https://github.com/gdsfactory/gdsfactory/discussions/424#discussioncomment-2867116).

**Reply by tvt173** [reply]
the other way you could do it would be to have the pdk config as a data (i.e. yaml) file in a set location in your project folder, and load it when starting gdsfactory. however, it would get nasty when you are trying to reference functions and things, and it would not be statically typed, like you desire. i think there's still a place for that approach (something like that would probably be a good replacement for `PDK.activate()`... instead just put a config file in your project and let gdsfactory _automatically activate for you_ based on the PDK you specify in your project config yaml). but for this particular issue, i prefer my suggestion above 🙂

**Comment by joamatab** [comment]
Hi!

just added this functionality of using strings registered in the PDK in https://github.com/gdsfactory/gdsfactory/pull/413/

and made a minor release gdsfactory 5.8.0

let me know what you think
Joaquin

**Comment by bradley-w-snyder** [comment]
Hello Joaquín,

I got a chance to look at it last night. I think I have some other issues
related to upgrading to Ubuntu 22.04 and also python 3.10 (make install is
no longer working, klayout was crashing and reinstalling required some
packages from outside the normal channels), but I don't think that's a
problem with gdsfactory.

Otherwise, it seems really cool and works well. I think there are still
some bugs though as I'm getting stuff randomly on the wrong layer (1/10,
1/12 and 68/0) where these layers are not defined in my PDK. I tracked down
one of the bugs to Component.show(), where 1/12 was hard-coded as an
argument. This should likely be "PORT" or a doc layer instead.

Best regards,
--Brad



Op do 26 mei 2022 om 18:46 schreef Joaquin Matres ***@***.***

**Reply by joamatab** [reply]
Yes, those layers are needed for connectivity checks

How about we add a flag to add them by default to the pdk, or raise a better error if they are not defined 

@tvt173 
@thomasdorch 
@Jan-David-Black 
@flaport

**Reply by tvt173** [reply]
hi @joamatab , better error sounds nice for sure. 

something like my [suggestion](https://github.com/gdsfactory/gdsfactory/discussions/409#discussioncomment-2862105) in the comment up above might help also. in that case, you could define the default required layers in some structure, and we could just add it into our PDK as-is, if desired, or add our own custom implementation if we want a different layer mapping (but still with a clear understanding of what those required layers are)

then if we have that, having a flag seems less important to me (i'd rather be specific and explicit on what I add to the PDK). but maybe it would be useful for getting started

**Comment by bradley-w-snyder** [comment]
P.S.

The random layers was a bug in my PDK where I was calling a component
definition function before the PDK finished initializing, so it was using
the default PDK and layers.

The other problems reminded me why I stopped using Linux for most things
almost 20 years ago, but they are fixed anyway. 😅

--Brad

Op wo 1 jun. 2022 09:13 schreef Bradley Snyder ***@***.***>:

**Comment by joamatab** [comment]
i added a validator in the PDK, that will raise an error if the layers needed for the connectivity checks are not defined

(DEVREC, PORT, PORTE)

https://github.com/gdsfactory/gdsfactory/pull/430

Another option is to provide default values for those layers instead of raising an error


what do you guys think?

**Comment by joamatab** [comment]
actually i think is cleaner to have some default layers added automatically to the PDK

https://github.com/gdsfactory/gdsfactory/pull/432

**Comment by bradley-w-snyder** [comment]
I prefer the validator. I don't like the idea of generating stuff on layers
by default that may not be expected. It should be explicitly set by every
PDK author if it is needed.

--Brad

Op do 2 jun. 2022 18:56 schreef Joaquin Matres ***@***.***>:

**Comment by joamatab** [comment]
Good point, yes, adding automatic layers can make the user confused

https://github.com/gdsfactory/gdsfactory/pull/433
